go-controller/cmd/ovnkube/ovnkube.go

```diff
@@ -230,12 +231,27 @@ func runOvnKube(ctx *cli.Context) error {
 		if runtime.GOOS == "windows" {
 			return fmt.Errorf("master nodes cannot be of OS type: Windows")
 		}
+		var ovnNBClient, ovnSBClient goovn.Client
+		var err error
+
+		if ovnNBClient, err = util.NewOVNNBClient(); err != nil {
```

Comment:
In my opinion and to try and keep loose coupling, it might be better to have a function such as `GetOVNNBClientInstance()` and invoke it where the OVN_NB and OVN_SB client instances are needed.? I noticed that function signatures had to be changed in a couple of places to pass the two client instances, hence the comment related to try and keep the coupling loose.

Reply:
We have gone back and forth here. the GetOVNNBClientInstance() assumes there is a global instance or at the very least needs to be a method on the controller or the controller needs to be passed in. In all such cases, something's gotta give. We need to pass in either the controller, or the instance of the client itself to any methods that are utility APIs such as `GetPortAddresses`, which don't have any receiver.

Reply:
@abhat I'd go for storing an instance in the controller, and then passing to any function that is not a struct function (eg GetPortAddresses())

Reply:
That is exactly what the current code does.

Reply:
How would you handle for below scenarios (I picked 2 there could be more)

```
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/cmd/ovn-kube-util/app/ovn-db-exporter.go#L272 invokes RunOVNSbctl which in turn is invoked by https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/cmd/ovn-kube-util/app/ovn-db-exporter.go#L352
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/node.go#L48 invokes RunOVNSbctl which in turn is invoked by https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/node.go#L162
```

Reply:
In those cases, you can create a separate instance of the NB/SB client inside the go-routine for `ovn-db-exporter.go`.